### PR TITLE
Add a template helper method to prefix relative URLs

### LINKED ISSUE #4871: Issue with relative URLs
### Affected version(s)

5.x

### Description

In #4851 we enabled the possibility to remove the `<base>` tag by outputting only absolute URLs. However, this does not work for any custom fields. For example:

```php
// contao/dca/tl_content.php
use Contao\CoreBundle\DataContainer\PaletteManipulator;

$GLOBALS['TL_DCA']['tl_content']['fields']['someLink'] = [
    'label' => ['Some link', 'Some link relating to the text.'],
    'inputType' => 'text',
    'eval' => ['rgxp' => 'url'],
    'sql' => ['type' => 'string', 'length' => 2048, 'default' => ''],
];

PaletteManipulator::create()
    ->addField('someLink', 'text_legend', PaletteManipulator::POSITION_APPEND)
    ->applyToPalette('text', 'tl_content')
;
```

```php
<!-- templates/ce_text.html5 -->
<?php $this->extend('ce_text'); ?>

<?php $this->block('content'); ?>
  <?php $this->parent(); ?>

  <?php if ($this->someLink): ?>
    <p class="someLink"><a href="<?= $this->someLink ?>">Some Link</a></p>
  <?php endif; ?>
<?php $this->endblock(); ?>
```

This will still output a relative URL and thus would not work if the `<base>` tag was removed.

In my original PR we discussed, that for any back end input field that can hold a URL a `save_callback` needs to be implemented so that the URL is always saved with `{{env::base}}` (or whichever insert tag is the correct one now) prefixed, if it's not an absolute URL (or insert tag) yet.

### PR DESCRIPTION
**Pull request by leofeyer**
Fixes #4871

I am open for a different method name.

@m-vo I am not sure if adding another dependency to the `ContaoExtension` class was the correct thing to do. If not, please point me in the right direction and I will adjust the PR accordingly.

### CODE REVIEW

**Comment by m-vo** [comment]
> @m-vo I am not sure if adding another dependency to the ContaoExtension class was the correct thing to do. If not, please point me in the right direction and I will adjust the PR accordingly.

I think we should create another runtime for this - then it will only get instantiated if needed and can have arbitrary dependencies:

```php
class UrlRuntime implements RuntimeExtensionInterface { // or whatever name fits
   public function __construct(private readonly RequestStack $requestStack) {}

    public function prefixUrl(string $url): string {
        // …
    }
}
```

And then reference it like the other runtimes.

**Comment by leofeyer** [comment]
Changed in b2b075581874a2f2ef26e5e8e7c2ba51f4963d40.

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Twig/Runtime/UrlRuntime.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    {
+    }
+
+    public function prefixUrl(string $url): string
+    {
+        if (!Validator::isRelativeUrl($url)) {
+            return $url;
+        }
+
+        return $this->requestStack->getMainRequest()?->getBasePath().'/'.$url;
```
Comment (m-vo) [inline core-bundle/src/Twig/Runtime/UrlRuntime.php:34]: Is the leading slash correct in case there is no main request?

Comment (ausi) [reply core-bundle/src/Twig/Runtime/UrlRuntime.php:34]: IMO Yes :)

**Review by m-vo — COMMENTED** [review]
Looks good to me!

Two small things:
1) What about adding a small test for the runtime? :grimacing: 
2) Are the CS changes to the tests intentional or a leftover of the initial version?

File: core-bundle/src/Twig/Runtime/UrlRuntime.php (hunk 1 above)
Comment (leofeyer) [reply core-bundle/src/Twig/Runtime/UrlRuntime.php:34]: Yes.

**Comment by leofeyer** [comment]
> 1. What about adding a small test for the runtime?

I don‘t think that makes sense. Both the validator and the `Request` class are tested, so we can safely assume that the method does what we expect it to do.

**Edit**: I have added tests anyway in 242457344362400c21df0421064ce3df6097dd7c. 😄 

> 2. Are the CS changes to the tests intentional or a leftover of the initial version?

The are leftover from the initial version. Do you want me to revert?

**Review by ausi — APPROVED** [review]

**Comment by m-vo** [comment]
:+1:

**Pull request merged**